### DISCUSSION TITLE
Field is missing with `name` attribute:

### DISCUSSION THREAD

**Discussion by rvision** [react-hook-form/react-hook-form#6568] [Q&A]
RHF version 6.15.8

https://codesandbox.io/s/react-hook-form-usefieldarray-forked-s98ix?file=/src/index.js

I am trying to setup a complex fieldsArray demo for which I am having serious issues, but I can't, here is what error I get when clicked on 'Add criteria' button (see console):

"Field is missing with `name` attribute: criterias"

I don't understand what does this mean in the first place....? Any help will be much appreciated

**Comment by bluebill1049** [comment]
I don't see that error message?
<img width="1055" alt="Screen Shot 2021-09-17 at 9 03 28 am" src="https://user-images.githubusercontent.com/10513364/133696400-52084c1d-d342-4eab-a923-efdf08582b55.png">

**Reply by rvision** [reply]
It happens when you click on 'Add criteria' button

**Reply by bluebill1049** [reply]
your input Field is missing with `name` attribute:  `criterias` try register('criterias'), however, look like you didn't even register a real input with useFieldArray, i am not sure it's going to work well with v6.

**Reply by rvision** [reply]
sorry, I am not following this at all...

which field is missing this attribute? do I have call register('criterias') even if this is going to be fieldsArray? it is already declared in defaultValues with useForm hook.

FYI: the code is not complete since I've stumbled to this error

**Reply by bluebill1049** [reply]
something is strange in your codesandbox, I couldn't reproduce it in this: https://codesandbox.io/s/react-hook-form-usefieldarray-forked-o20tu?file=/src/index.js this warning only gets through when an input gets registered without `name` attribute, also the key should be used `_key` as you renamed. Also, your input all missing defaultValue as well which is required in v6.

<img width="723" alt="Screen Shot 2021-09-17 at 6 55 33 pm" src="https://user-images.githubusercontent.com/10513364/133754831-c2bb6a42-a1ec-4e82-8aec-5c681eda3c01.png">

is any chance you can move to v7? as v7's useFieldArray is much better in general and has less restriction as well.

**Reply by rvision** [reply]
I've couldn't get normal values from .getValues() in any given path/handler, therefore I had to add my own key for items. It's probably going to be removed in final version and use item's field to mark fieldArray item for editing, but I cannot get this basic demo to work

I've  updated the CS demo:
- removed key - now it is generated by fieldsArray
- all default values are there and set to registered fields and controllers

still the same error when you click on 'Add criteria' button

regarding v7, I am not sure if it is doable, I don't know how many breaking changes are there and I will have to refactor existing code

**Reply by rvision** [reply]
I've updated the CSB with RHF 7 and added <DevTool />

https://codesandbox.io/s/react-hook-form-usefieldarray-forked-s98ix?file=/src/index.js

This is what I am trying to make:

![image](https://user-images.githubusercontent.com/1709545/133816549-08d8a53a-33ea-412d-a805-dd39cff8af72.png)

I already have semi-complex form with static fields and this is the only section that needs fieldsArray. Let me explain how this should work and how did I tried to accomplish this, especially if I am taking the wrong path here

1) I have read-only (display) rows and rows in edit mode: I've thought about having both of them rendered at the same time, but non-active one will be hidden by CSS. Part of the row schema is property 'inEditMode' which will be flag noting what to render, either display row or inline form. This has to be in the validation since user won't be allowed to submit form and continue if any field is inEditMode. When I wire on "Edit" button setValue(`${namePrefix}isInEditMode`, true, { shouldDirty: true,  shouldTouch: true }); it doesn't work, form doesn't re-render. How do I enforce this update? I've tried with Controller (see CSB), but that works only in one way and I am not sure if this is the proper way to do it

2) If I click on delete, in any situation id/key or the current fieldsArray row is wrong. Either if I try to get it via .getValues() or inline. How to fix this?

3) The form itself is not that complex in way that many re-renders will sacrifice performance. Is there any way to re-render whole component and get proper values in handlers when needed?

I updated RHF to v7, ran unit tests and nothing seems to be broken on UI, but will investigate later on when I finish this.

Thanks for any help

**Reply by bluebill1049** [reply]
This looks great. let me know if you got stuck or anything, v7 useFieldArray should be a lot nicer. (more features as well)

**Reply by rvision** [reply]
We have a misunderstanding here

All this I wrote is description of something that's not working.. I've tried bunch of variations of the above with no luck, that's what I am saying...

**Reply by bluebill1049** [reply]
> The form itself is not that complex in way that many re-renders will sacrifice performance. Is there any way to re-render whole component and get proper values in handlers when needed?

you can watch the entire form

**Reply by bluebill1049** [reply]
> If I click on delete, in any situation id/key or the current fieldsArray row is wrong. Either if I try to get it via .getValues() or inline. How to fix this?

if you want to use `id` or `key`, make sure you rename the `keyName` and use that renamed key value

`<div key={field.yourRenamed} />`

**Reply by bluebill1049** [reply]
> I have read-only (display) rows and rows in edit mode: I've thought about having both of them rendered at the same time, but non-active one will be hidden by CSS. Part of the row schema is property 'inEditMode' which will be flag noting what to render, either display row or inline form. This has to be in the validation since user won't be allowed to submit form and continue if any field is inEditMode. When I wire on "Edit" button setValue(${namePrefix}isInEditMode, true, { shouldDirty: true, shouldTouch: true }); it doesn't work, form doesn't re-render. How do I enforce this update? I've tried with Controller (see CSB), but that works only in one way and I am not sure if this is the proper way to do it

this should work by your description, can you try it in an isolated codesandbox to identity the issue?
